src/main/java/picard/analysis/directed/RnaSeqMetricsCollector.java

```diff
@@ -232,25 +228,60 @@ public void acceptRecord(SAMRecord rec) {
-                    ++metrics.CORRECT_STRAND_READS;
-                } else {
-                    ++metrics.INCORRECT_STRAND_READS;
+            if (overlapsExon && overlappingGenes.size() == 1) {
+                final Gene gene                           = overlappingGenes.iterator().next();
+                final boolean negativeTranscriptionStrand = gene.isNegativeStrand();
+                final boolean readOneOrUnpaired           = !rec.getReadPairedFlag() || rec.getFirstOfPairFlag();
+                final boolean negativeReadStrand          = rec.getReadNegativeStrandFlag();
+                if (strandSpecificity != StrandSpecificity.NONE) {
+                    final boolean readAndTranscriptStrandsAgree = negativeReadStrand == negativeTranscriptionStrand;
+                    final boolean firstReadExpectedToAgree      = strandSpecificity == StrandSpecificity.FIRST_READ_TRANSCRIPTION_STRAND;
+                    final boolean thisReadExpectedToAgree       = readOneOrUnpaired == firstReadExpectedToAgree;
+                    // If the read strand is the same as the strand of the transcript, and the end is the one that is supposed to agree,
+                    // then the strand specificity for this read is correct.
+                    // -- OR --
+                    // If the read strand is not the same as the strand of the transcript, and the end is not the one that is supposed
+                    // to agree, then the strand specificity for this read is correct.
+                    if (readAndTranscriptStrandsAgree == thisReadExpectedToAgree) {
+                        ++metrics.CORRECT_STRAND_READS;
+                    } else {
+                        ++metrics.INCORRECT_STRAND_READS;
+                    }
                 }
-            }
 
+                // Count templates only once rather than individual reads/records
+                if (readOneOrUnpaired) {
+                    // Check that for paired end reads they are in the proper orientation (FR) and that the entire
+                    // template is enclosed in the gene.
+                    final boolean properOrientation;
+                    final int leftMostAlignedBase, rightMostAlignedBase;
+                    if (rec.getReadPairedFlag()) {
+                        if (rec.getMateUnmappedFlag()) {
+                            properOrientation   = false; // mate is unmapped!
+                            leftMostAlignedBase = rightMostAlignedBase = 0;
+                        }
+                        else {
+                            properOrientation    = SamPairUtil.getPairOrientation(rec) == SamPairUtil.PairOrientation.FR;
+                            leftMostAlignedBase  = Math.min(rec.getAlignmentStart(), rec.getMateAlignmentStart());
+                            rightMostAlignedBase = Math.max(rec.getAlignmentEnd(), SAMUtils.getMateAlignmentEnd(rec)); // requires the mate cigar
```

Comment:
I think this is true for `FR` reads, but not `FF` reads.  Unfortunately, Picard/htsjdk computes insert size as 5' to 5' rather than left-most to right-most bases.